File: backend/internal/sanitize/sanitize_test.go
package sanitize

import (
	"testing"
)

func TestText(t *testing.T) {
	tests := []struct {
		name     string
		input    string
		expected string
	}{
		{
			name:     "strips script tags",
			input:    "<script>alert('xss')</script>Hello",
			expected: "Hello",
		},
		{
			name:     "strips style tags",
			input:    "<style>body{display:none}</style>Content",
			expected: "Content",
		},
		{
			name:     "strips HTML tags",
			input:    "<b>Bold</b> <i>Italic</i>",
			expected: "Bold Italic",
		},
		{
			name:     "strips event handlers",
			input:    `<img onerror="alert(1)" src="x">`,
			expected: "",
		},
		{
			name:     "handles empty string",
			input:    "",
			expected: "",
		},
		{
			name:     "normalizes whitespace",
			input:    "  hello   world  ",
			expected: "hello world",
		},
		{
			name:     "preserves normal text",
			input:    "Hello World",
			expected: "Hello World",
		},
		{
			name:     "decodes HTML entities",
			input:    "&lt;script&gt;",
			expected: "<script>",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := Text(tt.input)
			if result != tt.expected {
				t.Errorf("Text(%q) = %q, want %q", tt.input, result, tt.expected)
			}
		})
	}
}

func TestHTML(t *testing.T) {
	tests := []struct {
		name     string
		input    string
		expected string
	}{
		{
			name:     "escapes HTML tags",
			input:    "<script>",
			expected: "&lt;script&gt;",
		},
		{
			name:     "escapes quotes",
			input:    `"quoted"`,
			expected: "&#34;quoted&#34;",
		},
		{
			name:     "escapes ampersand",
			input:    "a & b",
			expected: "a &amp; b",
		},
		{
			name:     "handles empty string",
			input:    "",
			expected: "",
		},
		{
			name:     "preserves normal text",
			input:    "Hello World",
			expected: "Hello World",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := HTML(tt.input)
			if result != tt.expected {
				t.Errorf("HTML(%q) = %q, want %q", tt.input, result, tt.expected)
			}
		})
	}
}

func TestURL(t *testing.T) {
	tests := []struct {
		name     string
		input    string
		expected string
	}{
		{
			name:     "allows https URLs",
			input:    "https://example.com",
			expected: "https://example.com",
		},
		{
			name:     "allows http URLs",
			input:    "http://example.com",
			expected: "http://example.com",
		},
		{
			name:     "allows mailto URLs",
			input:    "mailto:test@example.com",
			expected: "mailto:test@example.com",
		},
		{
			name:     "allows relative URLs",
			input:    "/path/to/page",
			expected: "/path/to/page",
		},
		{
			name:     "blocks javascript URLs",
			input:    "javascript:alert('xss')",
			expected: "",
		},
		{
			name:     "blocks data URLs",
			input:    "data:text/html,<script>alert(1)</script>",
			expected: "",
		},
		{
			name:     "blocks vbscript URLs",
			input:    "vbscript:alert('xss')",
			expected: "",
		},
		{
			name:     "handles case variations",
			input:    "JAVASCRIPT:alert('xss')",
			expected: "",
		},
		{
			name:     "trims whitespace",
			input:    "  https://example.com  ",
			expected: "https://example.com",
		},
		{
			name:     "handles empty string",
			input:    "",
			expected: "",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := URL(tt.input)
			if result != tt.expected {
				t.Errorf("URL(%q) = %q, want %q", tt.input, result, tt.expected)
			}
		})
	}
}

func TestEmail(t *testing.T) {
	tests := []struct {
		name     string
		input    string
		expected string
	}{
		{
			name:     "accepts valid email",
			input:    "user@example.com",
			expected: "user@example.com",
		},
		{
			name:     "normalizes to lowercase",
			input:    "USER@EXAMPLE.COM",
			expected: "user@example.com",
		},
		{
			name:     "rejects invalid email - no @",
			input:    "not-an-email",
			expected: "",
		},
		{
			name:     "rejects invalid email - no domain",
			input:    "user@",
			expected: "",
		},
		{
			name:     "rejects invalid email - no local part",
			input:    "@example.com",
			expected: "",
		},
		{
			name:     "handles empty string",
			input:    "",
			expected: "",
		},
		{
			name:     "strips HTML from email",
			input:    "<script>alert(1)</script>user@example.com",
			expected: "user@example.com",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := Email(tt.input)
			if result != tt.expected {
				t.Errorf("Email(%q) = %q, want %q", tt.input, result, tt.expected)
			}
		})
	}
}

func TestFilename(t *testing.T) {
	tests := []struct {
		name     string
		input    string
		expected string
	}{
		{
			name:     "preserves valid filename",
			input:    "document.pdf",
			expected: "document.pdf",
		},
		{
			name:     "preserves hyphens and underscores",
			input:    "my-file_v2.txt",
			expected: "my-file_v2.txt",
		},
		{
			name:     "removes path traversal",
			input:    "../../../etc/passwd",
			expected: "etcpasswd",
		},
		{
			name:     "removes forward slashes",
			input:    "path/to/file.txt",
			expected: "pathtofile.txt",
		},
		{
			name:     "removes backslashes",
			input:    "path\\to\\file.txt",
			expected: "pathtofile.txt",
		},
		{
			name:     "replaces dangerous characters",
			input:    "file<script>.txt",
			expected: "file_script_.txt",
		},
		{
			name:     "handles empty string",
			input:    "",
			expected: "",
		},
		{
			name:     "handles dot only",
			input:    ".",
			expected: "unnamed",
		},
		{
			name:     "handles double dot",
			input:    "..",
			expected: "unnamed",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := Filename(tt.input)
			if result != tt.expected {
				t.Errorf("Filename(%q) = %q, want %q", tt.input, result, tt.expected)
			}
		})
	}
}

func TestSQLString(t *testing.T) {
	tests := []struct {
		name     string
		input    string
		expected bool
	}{
		{
			name:     "accepts normal text",
			input:    "Hello World",
			expected: true,
		},
		{
			name:     "accepts empty string",
			input:    "",
			expected: true,
		},
		{
			name:     "detects UNION SELECT",
			input:    "1 UNION SELECT * FROM users",
			expected: false,
		},
		{
			name:     "detects DROP TABLE",
			input:    "'; DROP TABLE users; --",
			expected: false,
		},
		{
			name:     "detects INSERT INTO",
			input:    "'; INSERT INTO users VALUES",
			expected: false,
		},
		{
			name:     "detects UPDATE SET",
			input:    "'; UPDATE users SET admin=1",
			expected: false,
		},
		{
			name:     "detects DELETE FROM",
			input:    "'; DELETE FROM users",
			expected: false,
		},
		{
			name:     "handles case variations",
			input:    "UNION select * from users",
			expected: false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := SQLString(tt.input)
			if result != tt.expected {
				t.Errorf("SQLString(%q) = %v, want %v", tt.input, result, tt.expected)
			}
		})
	}
}

func TestStripNullBytes(t *testing.T) {
	tests := []struct {
		name     string
		input    string
		expected string
	}{
		{
			name:     "removes null bytes",
			input:    "hello\x00world",
			expected: "helloworld",
		},
		{
			name:     "preserves normal text",
			input:    "hello world",
			expected: "hello world",
		},
		{
			name:     "handles multiple null bytes",
			input:    "\x00hello\x00\x00world\x00",
			expected: "helloworld",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := StripNullBytes(tt.input)
			if result != tt.expected {
				t.Errorf("StripNullBytes(%q) = %q, want %q", tt.input, result, tt.expected)
			}
		})
	}
}

func TestTruncateString(t *testing.T) {
	tests := []struct {
		name      string
		input     string
		maxLength int
		expected  string
	}{
		{
			name:      "truncates long string",
			input:     "Hello World",
			maxLength: 5,
			expected:  "Hello",
		},
		{
			name:      "preserves short string",
			input:     "Hi",
			maxLength: 5,
			expected:  "Hi",
		},
		{
			name:      "handles exact length",
			input:     "Hello",
			maxLength: 5,
			expected:  "Hello",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := TruncateString(tt.input, tt.maxLength)
			if result != tt.expected {
				t.Errorf("TruncateString(%q, %d) = %q, want %q", tt.input, tt.maxLength, result, tt.expected)
			}
		})
	}
}

func TestNormalizeName(t *testing.T) {
	tests := []struct {
		name     string
		input    string
		expected string
	}{
		{
			name:     "title cases name",
			input:    "john doe",
			expected: "John Doe",
		},
		{
			name:     "handles uppercase",
			input:    "JOHN DOE",
			expected: "John Doe",
		},
		{
			name:     "handles mixed case",
			input:    "jOHN dOE",
			expected: "John Doe",
		},
		{
			name:     "strips HTML",
			input:    "<b>John</b> Doe",
			expected: "John Doe",
		},
		{
			name:     "handles empty string",
			input:    "",
			expected: "",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := NormalizeName(tt.input)
			if result != tt.expected {
				t.Errorf("NormalizeName(%q) = %q, want %q", tt.input, result, tt.expected)
			}
		})
	}
}

func TestPassword(t *testing.T) {
	tests := []struct {
		name      string
		input     string
		minLength int
		maxLength int
		expected  string
	}{
		{
			name:      "accepts valid password",
			input:     "MySecurePass123!",
			minLength: 8,
			maxLength: 128,
			expected:  "",
		},
		{
			name:      "rejects short password",
			input:     "short",
			minLength: 8,
			maxLength: 128,
			expected:  "password too short",
		},
		{
			name:      "rejects long password",
			input:     "a" + string(make([]byte, 200)),
			minLength: 8,
			maxLength: 128,
			expected:  "password too long",
		},
		{
			name:      "rejects null bytes",
			input:     "password\x00injection",
			minLength: 8,
			maxLength: 128,
			expected:  "password contains invalid characters",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := Password(tt.input, tt.minLength, tt.maxLength)
			if result != tt.expected {
				t.Errorf("Password() = %q, want %q", result, tt.expected)
			}
		})
	}
}

File: frontend/app/lib/sanitize.test.ts
import { describe, expect, it } from "vitest";

import {
  escapeHtml,
  sanitizeBasicHtml,
  sanitizeEmail,
  sanitizeFilename,
  sanitizeObject,
  sanitizeRichHtml,
  sanitizeText,
  sanitizeUrl,
} from "./sanitize";

describe("sanitizeText", () => {
  it("should strip all HTML tags", () => {
    expect(sanitizeText("<script>alert('xss')</script>Hello")).toBe("Hello");
    expect(sanitizeText("<b>Bold</b> text")).toBe("Bold text");
    expect(sanitizeText("<img src=x onerror=alert(1)>")).toBe("");
  });

  it("should handle empty and null inputs", () => {
    expect(sanitizeText("")).toBe("");
    expect(sanitizeText(null)).toBe("");
    expect(sanitizeText(undefined)).toBe("");
  });

  it("should trim whitespace", () => {
    expect(sanitizeText("  hello  ")).toBe("hello");
  });

  it("should preserve normal text", () => {
    expect(sanitizeText("Hello World")).toBe("Hello World");
    expect(sanitizeText("user@example.com")).toBe("user@example.com");
  });
});

describe("sanitizeBasicHtml", () => {
  it("should allow basic formatting tags", () => {
    expect(sanitizeBasicHtml("<b>Bold</b>")).toBe("<b>Bold</b>");
    expect(sanitizeBasicHtml("<i>Italic</i>")).toBe("<i>Italic</i>");
    expect(sanitizeBasicHtml("<a href='https://example.com'>Link</a>")).toContain("<a");
  });

  it("should strip dangerous tags", () => {
    expect(sanitizeBasicHtml("<script>alert('xss')</script>")).toBe("");
    expect(sanitizeBasicHtml("<iframe src='evil.com'></iframe>")).toBe("");
    // Form and input tags are stripped, content is preserved
    expect(sanitizeBasicHtml("<form>content</form>")).not.toContain("<form");
  });

  it("should strip event handlers", () => {
    expect(sanitizeBasicHtml('<img onerror="alert(1)" src="x">')).not.toContain("onerror");
    expect(sanitizeBasicHtml('<div onclick="evil()">text</div>')).not.toContain("onclick");
  });
});

describe("sanitizeRichHtml", () => {
  it("should allow heading tags", () => {
    expect(sanitizeRichHtml("<h1>Title</h1>")).toBe("<h1>Title</h1>");
    expect(sanitizeRichHtml("<h2>Subtitle</h2>")).toBe("<h2>Subtitle</h2>");
  });

  it("should allow code blocks", () => {
    expect(sanitizeRichHtml("<pre><code>const x = 1;</code></pre>")).toContain("<code>");
  });

  it("should strip dangerous content", () => {
    expect(sanitizeRichHtml("<script>alert('xss')</script>")).toBe("");
    expect(sanitizeRichHtml("<style>body{display:none}</style>")).toBe("");
  });
});

describe("sanitizeUrl", () => {
  it("should allow safe URLs", () => {
    expect(sanitizeUrl("https://example.com")).toBe("https://example.com");
    expect(sanitizeUrl("http://example.com")).toBe("http://example.com");
    expect(sanitizeUrl("mailto:test@example.com")).toBe("mailto:test@example.com");
    expect(sanitizeUrl("/relative/path")).toBe("/relative/path");
  });

  it("should block dangerous protocols", () => {
    expect(sanitizeUrl("javascript:alert('xss')")).toBe("");
    expect(sanitizeUrl("data:text/html,<script>alert(1)</script>")).toBe("");
    expect(sanitizeUrl("vbscript:alert('xss')")).toBe("");
  });

  it("should handle case variations", () => {
    expect(sanitizeUrl("JAVASCRIPT:alert('xss')")).toBe("");
    expect(sanitizeUrl("JaVaScRiPt:alert('xss')")).toBe("");
  });

  it("should handle empty inputs", () => {
    expect(sanitizeUrl("")).toBe("");
    expect(sanitizeUrl(null)).toBe("");
    expect(sanitizeUrl(undefined)).toBe("");
  });

  it("should trim whitespace", () => {
    expect(sanitizeUrl("  https://example.com  ")).toBe("https://example.com");
  });
});

describe("sanitizeEmail", () => {
  it("should accept valid emails", () => {
    expect(sanitizeEmail("user@example.com")).toBe("user@example.com");
    expect(sanitizeEmail("USER@EXAMPLE.COM")).toBe("user@example.com");
  });

  it("should reject invalid emails", () => {
    expect(sanitizeEmail("not-an-email")).toBe("");
    expect(sanitizeEmail("@example.com")).toBe("");
    expect(sanitizeEmail("user@")).toBe("");
  });

  it("should strip HTML from email and validate result", () => {
    // After stripping HTML, if the resulting string is a valid email, it's kept
    expect(sanitizeEmail("<script>alert(1)</script>user@example.com")).toBe("user@example.com");
    // If stripping HTML leaves an invalid email, it returns empty
    expect(sanitizeEmail("<script>user@example.com</script>")).toBe("");
  });
});

describe("escapeHtml", () => {
  it("should escape HTML special characters", () => {
    expect(escapeHtml("<script>")).toBe("&lt;script&gt;");
    expect(escapeHtml('"quoted"')).toBe("&quot;quoted&quot;");
    expect(escapeHtml("a & b")).toBe("a &amp; b");
  });

  it("should handle empty inputs", () => {
    expect(escapeHtml("")).toBe("");
    expect(escapeHtml(null)).toBe("");
    expect(escapeHtml(undefined)).toBe("");
  });

  it("should preserve normal text", () => {
    expect(escapeHtml("Hello World")).toBe("Hello World");
  });
});

describe("sanitizeObject", () => {
  it("should sanitize string values", () => {
    const input = {
      name: "<script>alert(1)</script>John",
      email: "john@example.com",
    };
    const result = sanitizeObject(input);
    expect(result.name).toBe("John");
    expect(result.email).toBe("john@example.com");
  });

  it("should handle nested objects", () => {
    const input = {
      user: {
        name: "<b>John</b>",
        profile: {
          bio: "<script>evil()</script>Hello",
        },
      },
    };
    const result = sanitizeObject(input);
    expect(result.user.name).toBe("John");
    expect(result.user.profile.bio).toBe("Hello");
  });

  it("should handle arrays", () => {
    const input = {
      tags: ["<script>", "normal", "<b>bold</b>"],
    };
    const result = sanitizeObject(input);
    expect(result.tags).toEqual(["", "normal", "bold"]);
  });

  it("should preserve non-string values", () => {
    const input = {
      count: 42,
      active: true,
      data: null,
    };
    const result = sanitizeObject(input);
    expect(result.count).toBe(42);
    expect(result.active).toBe(true);
    expect(result.data).toBe(null);
  });
});

describe("sanitizeFilename", () => {
  it("should remove path traversal attempts", () => {
    expect(sanitizeFilename("../../../etc/passwd")).toBe("etcpasswd");
    expect(sanitizeFilename("..\\..\\windows\\system32")).toBe("windowssystem32");
  });

  it("should remove dangerous characters", () => {
    expect(sanitizeFilename("file<script>.txt")).toBe("file_script_.txt");
    expect(sanitizeFilename("file\x00.txt")).toBe("file.txt");
  });

  it("should handle empty inputs", () => {
    expect(sanitizeFilename("")).toBe("");
    expect(sanitizeFilename(null)).toBe("");
    expect(sanitizeFilename(".")).toBe("unnamed");
    expect(sanitizeFilename("..")).toBe("unnamed");
  });

  it("should preserve valid filenames", () => {
    expect(sanitizeFilename("document.pdf")).toBe("document.pdf");
    expect(sanitizeFilename("my-file_v2.txt")).toBe("my-file_v2.txt");
  });
});
